[Platform] Update musl modulemap and improve musl platform support.

al45tair opened the pull request:
We need to declare separate top-level modules for every header in the musl C library.  Note that our set-up here does require some tweaks to musl itself (specifically, the `alltypes.h` header needs some fairly serious massaging to make it module compatible; in particular, we need to pull out every type declaration into its own separate modularized header in `bits/types`).

Also update `Platform.swift` a little for more musl support.

rdar://123503615

al45tair committed:
[Platform] Update musl modulemap and improve musl platform support.

We need to declare separate top-level modules for every header in
the musl C library.  Note that our set-up here does require some
tweaks to musl itself (specifically, the `alltypes.h` header needs
some fairly serious massaging to make it module compatible; in
particular, we need to pull out every type declaration into its
own separate modularized header in `bits/types`).

Also update `Platform.swift` a little for more musl support.

rdar://123503615

al45tair commented:
This sits on top of #71839.

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
@swift-ci Please smoke test

ian-twilightcoder commented on stdlib/public/Platform/musl.modulemap:
```diff
@@ -2,110 +2,876 @@
+}
+
+module stdc_assert [system] {
+  config_macros NDEBUG
```
`config_macros` can't be used in submodules, only top level modules. Also config macros don't apply to textual headers.

ian-twilightcoder commented on stdlib/public/Platform/musl.modulemap:
```diff
@@ -2,110 +2,876 @@
+module stdc_assert [system] {
+  config_macros NDEBUG
+  textual header "assert.h"
+  export *
```
textual headers don't export * because they don't build with their module and thus don't import anything to be exported. Another issue with textual headers is they need to not declare anything, otherwise the declarations will be written to multiple modules, with potentially different definitions. assert.h is particularly snarly though, it has to define `assert`, but anything that uses `assert` in the OS/SDK needs to `undef assert` so that its module doesn't bake in a declaration for `assert` that's potentially incompatible with other modules.

al45tair replied:
There's a limit to the amount of change I can (easily) make to the musl headers, so to the extent that this problem exists it's going to have to remain a problem for now; even if we can upstream changes to resolve this, experience so far suggests it's quite difficult to get and sustain the attention of musl's (sole) maintainer, and he's the only one who actually merges changes as far as I can tell.

al45tair replied:
(The changes I've made, for instance to put all the types in separate headers in `bits`, are done automatically by a script.)

ian-twilightcoder replied:
Yeah, sure, but we can still drop the `export *` since it doesn't do anything. 😄

ian-twilightcoder commented on stdlib/public/Platform/musl.modulemap:
```diff
@@ -2,110 +2,876 @@
+  export *
+}
+
+// POSIX
```
I don't know how we want to organize the rest of these. One module per header can start to cause performance issues, but having too many headers in a module can also cause performance issues. Is there some kind of logical grouping we might be able to use like filesystem, networking, etc?

al45tair replied:
I _guess_ these are less likely than the C library headers to suffer from the circularity issues created by the builtin modules and the C++ library, so we probably could group them. I just broke them all out for now.

al45tair replied:
I started looking at the Android module map that's being worked on to see how things were grouped there. I'm wondering if we should try to make things match. Not sure that's a goal right now, but it's an interesting question.

ian-twilightcoder commented:
Thank you for doing this one!

ian-twilightcoder commented on stdlib/public/Platform/musl.modulemap:
```diff
@@ -2,110 +2,876 @@
+  }
+}
+
+module stdc_assert [system] {
```
If we do it this way, then we can drop all of the platform checks and just `import stdlib_h` etc from Swift once we get the rest of the module maps converted.

al45tair replied:
Interesting. I don't think we want the Swift interface to the C library to be things like `import stdlib_h` though. That seems like the wrong approach. `import Libc` would be my personal preference, and maybe `import Posix` and `import LibcExt` also, the idea being that `Libc` will get you a standard set of Libc declarations (ones from the C standard), `Posix` will get you a standard set of Posix declarations (possibly even on Windows we could provide some subset there), and `LibcExt` would get you any platform extensions.

I have some ideas about how to get there; it's a little more complicated than just dropping module maps in the right places, obviously, because the entire point is that the _Swift_ interface for `Libc` and `Posix` would be standardised (e.g. nullability annotations on the C declarations would be the same, regardless of what the underlying C library has done).

al45tair replied:
As regards the first issue, that's presumably a side effect of not having `-fbuiltin-headers-in-system-modules`? With that flag, it'll find the correct header, but without it we're going to need your workaround.

ian-twilightcoder replied:
No `-fbuiltin-headers-in-system-modules` isn't involved. `header` declarations always resolve relative to the module map file and do not use header search paths. Otherwise the OS/SDK module would accidentally cover the libc++ or clang stdint.h instead of its own stdint.h. module maps can't be expected to know about other headers with the same name in different directories, or how those layer with everything else.

I would still underscore the modules for the real headers so that those don't show up in autocomplete.

Personally I prefer having cover modules for each of the headers because they aren't that related to each other. e.g. I might need something out of string.h or time.h but I don't want to pull in anything from signal.h or stdlib.h. There's also the performance concern: a `Libc` module will end up containing a lot of headers, e.g. on Apple platforms just stdlib.h pulls in 75 headers. That's a lot of headers to stat and open when all someone wanted was e.g. `size_t`, and it gets real expensive when NFS gets involved. POSIX is much worse, and probably shouldn't be used all that often from ObjC or Swift anyway.

That's my opinion at least. I get the convenience of having to import only one or two things and you get everything you need, but you also get a huge pile of junk in your autocomplete that you don't need, and it slows down builds too.

al45tair replied:
> If we do it this way, then we can drop all of the platform checks and just `import stdlib_h` etc from Swift once we get the rest of the module maps converted.

We'd have to fix a few more things before that would work, particularly relating to the Clang Importer (it would need to guarantee to import the C library in a uniform manner in spite of C-source-compatible differences in declarations on different platforms).

al45tair replied:
I've adopted the same approach (the `_modules/<modulename>.h` one, I mean) in the musl modulemap, as that seems like a neat idea.

ian-twilightcoder commented:
We also need to change the clang importer to not pass `-fbuiltin-headers-in-system-modules ` when using musl

al45tair commented on stdlib/public/Platform/musl.modulemap:
```diff
@@ -2,110 +2,876 @@
-module SwiftMusl [system] {
-  header "stdc-predef.h"
+module _musl [system] {
+  config_macros _GNU_SOURCE, _BSD_SOURCE, _XOPEN_SOURCE, _POSIX_SOURCE, _POSIX_C_SOURCE, _LARGEFILE64_SOURCE, _ALL_SOURCE, _DEFAULT_SOURCE, __STRICT_ANSI__
```
That isn't a problem here, because bridging headers are an Xcode thing, and this isn't anything to do with Xcode. Also, the config macro declarations have actually shown up a number of build problems in Swift packages (which I've filed PRs for), so they're certainly useful.

I _think_ I'm right in saying that we _can_ use `config_macros` on Linux, *BSD, Musl and Windows. The place we can't is Darwin because of the bridging header issue.

al45tair replied:
Yes. ObjC interop doesn't exist on non-Darwin, and even on Darwin if I were writing a program and not using Xcode, I doubt I'd use that flag — I'd probably make an explicit module map instead.

al45tair committed:
[Platform] Updated after review comments.

Added the _modules workaround to make the compiler pull in the correct
module when someone does `import stdc_assert` (for instance).

Fix various niggles.

rdar://123503615

al45tair commented:
> We also need to change the clang importer to not pass `-fbuiltin-headers-in-system-modules ` when using musl

That is [in this changeset](https://github.com/apple/swift/pull/71844/commits/51553a616b4f4bef205cedd4d22b9b8d9c8c7a32) as part of #71844.

al45tair commented:
@swift-ci Please smoke test

ian-twilightcoder reviewed (changes requested):
I really wouldn't do the one module per header thing for the non-libc ones. That's proving to be not very good for performance if more than a small handful get used. But mostly I'd like to rename the libc ones to be consistent with other modules.

al45tair committed:
[Platform] Rename modules for C library headers to match Darwin.

The Darwin headers have a naming convention for these modules; we should
match it.

rdar://123503615

al45tair commented:
@swift-ci Please smoke test

al45tair commented on stdlib/public/Platform/musl.modulemap:
```diff
@@ -2,110 +2,1002 @@
+// ensures that e.g. `import stdc_float` will result in pulling in the Clang
+// or libc++ versions as expected.
+
+module _stdc_assert [system] {
```
I've updated the names of the C library headers to match.

al45tair commented:
@swift-ci Please smoke test
